Add mod.ini import support

asarium opened the pull request:
This adds an action to import a legacy mod into the Knossos system which
performs the most annoying steps automatically. This will try to convert
the old style dependency management to the new style by using some
hard-coded conversion values. Currently, that table only contains the
various MediaVP versions.

This will also extract the individual VPs and create packages for each
of them. That is currently done on the main thread which means that the
GUI is blocked while the VPs are being extracted. I don't know how the
task system of Knossos works yet so I chose to implement the basic
version first.

ngld commented:
This PR needs some refactoring before I can merge it.

* The extraction should happen in a `Task` (as you already mentioned).
* it seems like a lot of the code is duplicated:
  * The new popup mode seems very similar to the create mod popup and should probably share the same code.
  * The VP extraction logic is now in two places in the code (both in `InstallTask` and your new code). This should be moved into its own function or into `VpReader`
  * `importIniMod` and `createMod` seem to share a lot of code.
* Your code shouldn't import `PyQt5` directly to maintain consistency with the existing code.
* Why does `parseIniMod` return the whole mod structure if only the ID and title are used?
* This one's more of a nitpick: Most of the existing code uses printf-style formatting of strings ('%s' % var) so it'd be nice if you used that instead of `format()`.

All existing tasks are declared in `tasks.py` and either inherit from `progress.Task` or `progess.MultistepTask`. A task can be triggered/run by passing it to `run_task`: `run_task(TaskName())`.

To give a short overview of how `Task` works: The constructor is usually run on the main thread and calls `self.add_work([...])` to add work items. For each work item `work` is called in a background thread (Multiple calls will happen at the same time. if you want to limit how many threads are running this task at the same time, you need to set `self._threads = <number>`).

Once the task is finished the `done` signal is emitted which is usually connected to the `finish` method which then runs on the main thread.
If `work` called `post(result)`, this method can call `self.get_result()` to get a list of results.

In this case you could pass a list of VP files to `add_work` and extract each VP file in `work`. If an error occurs, you'd need to set an error flag (like `self._error = True`) and check that in the `finish` method to display an error message.

asarium commented:
I moved the VP extraction into a task and also extracted the VP extraction code into a single function. The handling of the `done` signal may be a bit creative but it makes the code more modular since the VP extraction task does not need to handle creating the packages and saving the mod.

I merged the mod creation and import popups and also restricted what kind of mods can be created with the import function (mod.ini never supported engines or tools and there are few enough TCs that handling that manually shouldn't be too much of a problem).

`parseIniMod` returns the entire structure since the code might need more data at some point and it was the easiest way of handling that task.

I changed `format()` to the `% ...` syntax.

ngld commented on knossos/web.py:
```diff
@@ -1141,6 +1144,111 @@ def testVoice(self, voice, volume, text):
+            mod.save()
+
+            center.installed.add_mod(mod)
+            mod.update_mod_flag()
```
Why did you call this after `mod.save()`?

asarium replied:
Because that function needed a reference to the repo and the `createMod` function called `save()` before adding the mod so I just kept that...

Anyway, I moved the save call below the `update_mod_flag` call.

ngld replied:
`save` has to be called after `update_mod_flag` otherwise the changes it just made will be lost on the next restart.

ngld commented on knossos/web.py:
```diff
@@ -26,6 +26,9 @@
 from datetime import datetime
+
+from knossos.repo import IniMod
+from knossos.tasks import VpExtractionTask
```
`repo` and `tasks` are already imported. Just use `repo.IniMod` and `tasks.VpExtractionTask` below instead of adding new imports.

asarium replied:
Fixed.

ngld commented on knossos/util.py:
```diff
@@ -34,6 +34,7 @@
 from threading import Condition, Event, Thread
 from collections import deque
 
+from knossos.vplib import VpReader
```
It'd be nicer if this line used relative imports like all the other `knossos` imports. (`from .vplib import VpReader`)

asarium replied:
Fixed.

ngld commented:
LGTM. I'll merge it as soon as I can.
